[vscode] 7/n listeners and .env

Ankush-lastmile opened the pull request:
[vscode] 7/n listeners and .env

Summary:

When the .env file path is set, vscode will invoke api/set_env on each AIConfig server.

Defines a listener to a newly introduced setting for the env_file_path.

This ensures that users do not need to restart vscode to pickup env changes.

Test Plan:

1. set env file path
2. changes picked up
3. changes picked up on restart
4. Changes picked up when manually removing vscode setting and re-setting env file path and api key

https://github.com/lastmile-ai/aiconfig/assets/141073967/07ac1d43-28ef-439f-98e7-94a4e73e11c9

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/1390).
* #1398
* __->__ #1390
* #1387
* #1389

Ankush-lastmile commented on vscode-extension/src/util.ts:
```diff
@@ -98,6 +102,15 @@ export async function updateServerState(
+    }
+  );
+}
+
```
for anyone unaware, 
`[ENV_FILE_PATH]` is syntax equivalent to resolving the value of `ENV_FILE_PATH` as the key for the js object.

rholinshead commented on vscode-extension/package.json:
```diff
@@ -128,6 +128,11 @@
+        "vscode-aiconfig.env_file_path": {
+          "type": "string",
+          "default": "",
+          "description": "env_file_path"
```
Description should be clearer on what this is for

Ankush-lastmile replied:
Updated to "File path to your .env file containing your API keys"

rholinshead commented on vscode-extension/src/util.ts:
```diff
@@ -383,6 +394,12 @@ export async function setupEnvironmentVariables(
     );
   }
+
+  // Update Server Env FLow
```
Do we technically still need `lowestCommonWorkspacePath` validation above now, or can .env path be anywhere?

Ankush-lastmile replied:
Not needed after this diff, will remove

rholinshead commented on vscode-extension/src/util.ts:
```diff
@@ -383,6 +394,12 @@ export async function setupEnvironmentVariables(
+  // Update Server Env FLow
+  // Set the .env file path in the settings
+  // vscode Extension has a listener for changes defined at activation.
+  const config = vscode.workspace.getConfiguration(EXTENSION_NAME);
```
Can you give some context on why we need to persist this? This would result in this persisted .env file being used for a given config instead of any .env file colocated in the folder with the config, right?

Ankush-lastmile replied:
>This would result in this persisted .env file being used for a given config instead of any .env file colocated in the folder with the config, right?

Correct. With the two following nuances
1.  Setting the env path would be stored in either the workspace or global workspace vscode settings depending on where you are.
2. A potential .env file colocated with the config would work IF the setting from 1 is not set, since `load_dotenv` will pull the the closest env file.

The only other alternative I see is to _not_ store `env_file_path` as an extension setting, and ONLY update the currently open editor file. However, I'm not sure how I feel about that vs. this way to persist and use the env_file_path for all aiconfigs in workspace.

rholinshead commented on vscode-extension/src/editor_server/editorServer.ts:
```diff
@@ -73,6 +74,8 @@ export class EditorServer {
       const modelRegistryPathArgs = modelRegistryPath
         ? ["--parsers-module-path", modelRegistryPath]
         : [];
+      const filePath = config.get(ENV_FILE_PATH) as string;
```
oh, one last thing - should this be `as string | undefined`?

Ankush-lastmile replied:
same as below

rholinshead commented on vscode-extension/src/extension.ts:
```diff
@@ -208,6 +210,25 @@ export async function activate(context: vscode.ExtensionContext) {
+  context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(async (event) => {
+    if (event.affectsConfiguration("vscode-aiconfig")) {
+      // Get new env
+      const envPath = vscode.workspace.getConfiguration("vscode-aiconfig").get(ENV_FILE_PATH) as string;
```
Same here, unless deleting it makes its value empty string instead of undefined

rholinshead commented on vscode-extension/src/extension.ts:
```diff
@@ -208,6 +210,25 @@ export async function activate(context: vscode.ExtensionContext) {
   );
+
+  // Handle changes to the .env path
+  context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(async (event) => {
```
no need for async

rholinshead commented on vscode-extension/src/extension.ts:
```diff
@@ -208,6 +210,25 @@ export async function activate(context: vscode.ExtensionContext) {
+        aiconfigEditorManager.getRegisteredEditors()
+      );
+      if (editors.length > 0) {
+        editors.forEach(async (editor) => {
```
no need for async if not awaiting

rholinshead reviewed (approved):
Just a few small things

Ankush-lastmile commented:
- added an await statement when invoking updateServerEnv
- modified event handler to execute only if the extension setting for env_file_path is modified
